src/iop/prover.rs

```diff
@@ -0,0 +1,833 @@
+            // with
+            //      M(alpha, X) = Sum_{z in H} L(alpha,z) * M(z,X)
+            for (r, row) in matrix.iter().enumerate() {
+                for (coeff, c) in row.iter() {
+                    let index = domain_h
+                        .reindex_by_subdomain(input_domain.size(), *c)
+                        .map_err(|e| Error::Other(e.to_string()))?;
+                    t_evals_on_h[index] += *eta * coeff * l_x_alpha_on_h[r];
+                }
+            }
+        }
+        Ok(t_evals_on_h)
+        // Ok(EvaluationsOnDomain::from_vec_and_domain(t_evals_on_h, domain_h).interpolate())
+    }
+
+    /// Returns the ratio of the sizes of `domain` and `subdomain` or an Error if
+    /// `subdomain` is not a subdomain of `domain`.
+    fn get_subdomain_step(
+        domain: &Box<dyn EvaluationDomain<F>>,
+        subdomain: &Box<dyn EvaluationDomain<F>>,
+    ) -> Result<usize, Error> {
+        if domain.size() % subdomain.size() != 0 {
+            Err(Error::Other(
+                "domain size not divisible by subdomain size".to_owned(),
+            ))?
+        }
+        let step = domain.size() / subdomain.size();
+        if subdomain.group_gen() != domain.group_gen().pow(&[step as u64]) {
+            Err(Error::Other(
+                "domain and subdomain have inconsistent generators".to_owned(),
+            ))?
+        }
+        Ok(step)
+    }
+
+    /// Output the number of oracles sent by the prover in the first round.
+    pub fn prover_num_first_round_oracles() -> usize {
+        3
+    }
+
+    /// Prover second round of the algebraic oracle proof, the "outer sumcheck" that
+    /// results from batching and reducing the R1CS identities.
+    /// Determines the oracles for `T(alpha, X)`, `U_1(X)` and `h_1(X)`.
+    pub fn prover_second_round<'a, R: RngCore>(
+        ver_message: &VerifierFirstMsg<F>,
+        mut state: ProverState<'a, F>,
+        zk: bool,
+        rng: &mut R,
+    ) -> Result<(ProverSecondOracles<F>, ProverState<'a, F>), Error> {
+        let round_time = start_timer!(|| "IOP::Prover::SecondRound");
+
+        let domain_h = &state.domain_h;
+
+        let alpha = ver_message.alpha;
+        let (eta_a, eta_b, eta_c) = ver_message.get_etas();
+        // In the following we exploit the fact that eta_a == 1 to avoid unnecessary
+        // multiplications.
+        assert_eq!(eta_a, F::one());
+
+        let summed_y_m_poly_time = start_timer!(|| "Compute y_m poly");
+        let (y_a_poly, y_b_poly) = match state.my_polys {
+            Some(ref v) => v,
+            None => return Err(Error::Other("mz_polys are empty".to_owned())),
+        };
+
+        // Performed via FFT using a domain of size = deg(y_a) + deg(y_b) + 1
+        // TODO: check if the `*` is really doing FFT here, and if so
+        // why we do the FFT product for the outer_poly below manually
+        let y_c_poly = y_a_poly.polynomial() * y_b_poly.polynomial();
```

Comment:
Yes it uses FFT

Reply:
outer_poly is computed in the same way. The evaluations of the polys over the b FFT domain are multiplied toghether.

Reply:
Agree. Comment removed